libraries/AP_LQR_Control/AP_LQR_Control.cpp

```diff
@@ -0,0 +1,413 @@
+
+    _crosstrack_error=- _crosstrack_error;
+    
+    //set minimum turn radius equal to twice the ground speed
+    float min_turn_rad = 2*groundSpeed;
+    
+    //limit maximum lateral acceleration and set default if zero
+    if((_max_latacc > (100*groundSpeed*groundSpeed/min_turn_rad)) || (_max_latacc ==0))
+    { 
+        _max_latacc = 100*groundSpeed*groundSpeed/min_turn_rad;
+    }
+    
+    //Caluclate adaptive gains
+    float q1= sqrtf(float((_max_xtrack*0.01)/(fabs((_max_xtrack*0.01)-_crosstrack_error))));
+    
+    //Calculate the approach velocity towards path
+    float si = RadiansToCentiDegrees(get_yaw_rad())*0.01;
+    float si_p = (get_bearing_cd(prev_WP,next_WP))*0.01;
+    float temp_sin=sinf(radians(si - si_p));
+    float v_d= groundSpeed * temp_sin;
+    
+    //Compute lateral acceleration based on current state and adaptive gains
+    float u =  - (((_xtrack_fac*0.01)*q1*_crosstrack_error)+(sqrtf((float)((_q2_val*0.01)+(2*q1)))*(_vel_fac*0.01)*v_d));
+    
+    //Limit calculated lateral acceleration to maximum values
+    if (fabs(u) > (_max_latacc*0.01))
+    {
+        if(u > 0)
+            u = _max_latacc*0.01;
+        if(u < 0)
+            u = - (_max_latacc*0.01);
+    }
+    _latAccDem= u;
+
+    
+    // Waypoint capture status is always false during waypoint following
+    _WPcircle = false;
+
+    
+
+    _data_is_stale = false; // status are correctly updated with current waypoint data
+}
+
+// update L1 control for loitering
+void AP_LQR_Control::update_loiter(const struct Location &center_WP, float radius, int8_t loiter_direction)
+{
+    struct Location _current_loc;
+    Vector2f _groundspeed_vector = _ahrs.groundspeed_vector();
+
+    // scale loiter radius with square of EAS2TAS to allow us to stay
+    // stable at high altitude
+    radius = loiter_radius(radius);
+    float groundSpeed=_groundspeed_vector.length();
+    Vector2f location_difference=location_diff(center_WP, _current_loc);
+    
+    //Calculate rate change of heading of path
+    float si_p_dot=groundSpeed/location_difference.length();
+    _crosstrack_error = location_difference.length() - radius;
+    
+    //Set minimum turn radius equal to twice the groundspeed
+    float min_turn_rad = 2*groundSpeed;
+    
+    //limit maximum lateral acceleration and set default if zero
+    if((_max_latacc > (100*groundSpeed*groundSpeed/min_turn_rad)) || (_max_latacc ==0))
+    { 
+        _max_latacc = 100*groundSpeed*groundSpeed/min_turn_rad;
+    }
+    
+    float u =0;
+    
+    // update _target_bearing_cd
+    _target_bearing_cd = get_bearing_cd(_current_loc, center_WP);
+    //Compute adaptive gains
+    float q1= sqrtf((float)((_max_xtrack*0.01)/(fabs((_max_xtrack*0.01)-_crosstrack_error))));
+    //Compute velocity of approach towards desired path
+    float si = RadiansToCentiDegrees(get_yaw_rad())*0.01;
+    // check if vehicle is not very far from the desired circular path
+    if (_crosstrack_error < (min_turning_radius))
```

Comment:
min_turning_radius not defined

Reply:
Is setting minimum turning radius twice the groundspeed a good way? Or should it be given as a parameter?

Reply:
Might be good to use WP_RADIUS, but for this, better to ask @magicrub and @WickedShell

Reply:
I prefer not to give users control over such factors. It is also being used for calculating maximum allowable lateral acceleration. But I believe I am the least experienced person in the field. I am open to suggestions.

Reply:
Turning radius (when tuned properly) is a function of bank angle and airspeed (LIM_ROLL_CD and TRIM_ARSPD_CM). That relationship is the following: R=(airspeed^2)/(9.81*tan(bank_angle)).

Reply:
@Naterater Thanks. Also, what about the rudder mixing factor? It is also supposed to effect the minimum turning radius. Any estimate for its effect?